app/controllers/dashboard_controller.rb

```diff
@@ -16,7 +16,13 @@ def index
 
   def show
     if @widget.present? && @widget.user == User.current
-      render(:partial => @widget.template, :locals => @widget.data)
+      cache_key = "#{@widget.cache_key}-#{@widget.user.id}/search-#{params[:search]}"
+      from_cache = Rails.cache.fetch(cache_key)
+      if from_cache
+        render html: from_cache
+      else
+        Rails.cache.write(cache_key, render(:partial => @widget.template, :locals => @widget.data), expires_in: Setting[:widget_cache])
+      end
```

Comment:
I could not figure out how to use `Rails.cache.fetch() do end` block because Rails stack raises exception on multiple `render` calls.